### DISCUSSION TITLE
Add automatic window scaling based on OS screen scale

### DISCUSSION THREAD

**Discussion by passivestar** [godotengine/godot-proposals#9825] [Editor]
The default `1152x648` was probably chosen with fullhd in mind, but on high res displays it's way too tiny, impossible to use for any kind of testing (this is a laptop screen):

<img width="1728" alt="Screenshot 2024-05-26 at 14 38 49" src="https://github.com/godotengine/godot-proposals/assets/60579014/a0f605de-a81f-4b07-8d6c-880dd958b99f"><br>

Whenever I create a project I have to change it. It would be nice if godot set it to something more usable depending on the screen resolution. Since it's used in exported projects too perhaps it would make sense for godot to detect OS screen scale if possible and use `1152x648` as a baseline resolution at 100% screen scale

**Comment by AThousandShips** [comment]
Since this is used for the exported project it shouldn't be based on the size of the display someone is developing on, should this just be defined when creating the project? Otherwise it'd be updated for each person working on the project, creating noise in the diff, what if you and your fellow developers have different screen sizes? What if you have different screen sizes on your laptop Vs your desktop?

I don't think it is useful to make the default size based on the screen size of the editor, it's easy to change if you want something else but it'd also annoy to have to change it if you don't agree with the detected size if you're targeting something specific

**Reply by passivestar** [reply]
Sorry it was probably unclear from my post but what I was suggesting is to keep `1152x648` in the project settings as a "base resolution" and scale the actual window resolution _at runtime_ if godot can detect screen scale from the OS:

<img width="450" alt="image" src="https://github.com/godotengine/godot-proposals/assets/60579014/d5216fed-4e7a-4b9f-b96a-5cfe44a7dede">

<img width="450" alt="image" src="https://github.com/godotengine/godot-proposals/assets/60579014/a734aaf8-8d7b-48be-adc7-9999c3012e02">

This scaling can be made opt-in (configurable in project settings)

As far as I understand apps don't necessarily always use absolute pixel values anymore because high ppi displays need to be supported. In apple's case when "retina" happened the OS just multiplied pixel values from third party apps by 2 automatically instead of making everything 2 times smaller 🙂

**Reply by AThousandShips** [reply]
My bad that was unclear given the:
> It would be nice if godot set it to something more usable depending on the screen resolution.

If the sentence following it is supposed to be the suggestion and not an alternative I think it should be clearer (especially with the title also reinforcing that assumption)

**Reply by AThousandShips** [reply]
But to me it makes little sense to have a windowed game open that's dependent on the size of the screen, as opposed to either a reasonable base size that works for *most* targets (i.e. a good baseline for designing UI and view size etc.) or fullscreen

I can't really think of any other app or game that I've used that works like that, i.e. takes up a certain portion of the screen automatically (without using desktop features)

I'd say the default project size is a tool to use to establish a baseline for what you're developing for, to get a sense of how to design your UI and game world so it doesn't feel too cramped on some designated base size or minimum supported size

**Reply by passivestar** [reply]
Renamed to better reflect the proposal

**Reply by passivestar** [reply]
@AThousandShips Maybe I misunderstand or don't know if it's already somehow supported but lets say you're making a desktop app to make music/sfx or an image editor of some sort and you want it to start windowed. Or perhaps it's a file conversion utility that has so few buttons in it that it doesn't even make sense for it to support maximized mode. To make the window not super tiny on a 4k laptop you need to use OS screen scale factor, no?

You can't make it work for _most_ targets because the PPI difference between a fullhd 16'' display and a 4k 16'' display is very big. Godot's default project window size is a good example, I don't have a ruler but I'd say it's about 15cm wide on MBP16, it's tiny

**Reply by AThousandShips** [reply]
You can always enable these features as well, but the proposed feature here is to have a *windowed* app running that's a certain fraction of the screen size, can you provide examples of that for the main window of any app?

For a *popup* you can use the dedicated methods for that

But I don't see this feature being useful for *most* of the things you'd develop in Godot, which is really what should be the governing thing for how we develop it. So this feels very niche, most of the time you'd either have a fixed window or full screen I'd say, for the reasons outlined above

**Reply by AThousandShips** [reply]
But for the edge case of having the *main* window be a proportional popup we could add a setting in the project settings, a mode perhaps along with `Windowed`, taking some proportional factor, i.e. like [`popup_centered_ratio`](https://docs.godotengine.org/en/stable/classes/class_window.html#class-window-method-popup-centered-ratio)

**Reply by passivestar** [reply]
> can you provide examples of that for the main window of any app?

Godot editor would be a good example. It automatically detects the screen scale, and thanks to that when the project manager opens it's not super tiny. I don't know if you consider the project manager a popup, but it's a good user experience when godot project manager window has the same consistent physical size regardless of what laptop I open it on. Seems like it would make sense for the exported apps to have that too

**Reply by AThousandShips** [reply]
That can be done with scripting, though the project manager defaults to a particular size not a fraction of the window size (though the popups do)

This would be useful as an opt-in alternative to other window modes, or allow percentages for the size, possibly with some way to make one proportional to the other size perhaps, to keep a ratio even if the screen is a different format (for example for wide screens), or pick the smallest size (this all should be possible with aspect ratios etc. as well)

**Comment by Calinou** [comment]
Thanks for the proposal! Consolidating in https://github.com/godotengine/godot-proposals/issues/7968.
